### ISSUE TITLE
Warning -- threading._dangling was modified by test_asyncio on FreeBSD 9 and 10

### ISSUE THREAD

**Issue by vstinner** [python/cpython#74466]
BPO | [30280](https://bugs.python.org/issue30280)
--- | :---
Nosy | @vstinner, @1st1, @koobs, @zhangyangyu
PRs | <li>python/cpython#2501</li><li>python/cpython#2511</li><li>python/cpython#2512</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-06-30.15:22:29.733>
created_at = <Date 2017-05-05.09:08:35.083>
labels = ['3.7', 'tests', 'expert-asyncio']
title = 'Warning -- threading._dangling was modified by test_asyncio on FreeBSD 9 and 10'
updated_at = <Date 2017-06-30.15:22:29.732>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-06-30.15:22:29.732>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-06-30.15:22:29.733>
closer = 'vstinner'
components = ['Tests', 'asyncio']
creation = <Date 2017-05-05.09:08:35.083>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 30280
keywords = []
message_count = 7.0
messages = ['293082', '293325', '293683', '297366', '297367', '297416', '297417']
nosy_count = 4.0
nosy_names = ['vstinner', 'yselivanov', 'koobs', 'xiang.zhang']
pr_nums = ['2501', '2511', '2512']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue30280'
versions = ['Python 3.5', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by vstinner** [comment]
On the FreeBSD 9 buildbot, I saw this warning:

Warning -- threading.\_dangling was modified by test_asyncio

http://buildbot.python.org/all/builders/AMD64%20FreeBSD%209.x%203.6/builds/124/steps/test/logs/warnings%20%281%29

http://buildbot.python.org/all/builders/AMD64%20FreeBSD%209.x%203.5/builds/78/steps/test/logs/warnings%20%281%29

**Comment by vstinner** [comment]
Warning also seen on FreeBSD 10:
http://buildbot.python.org/all/builders/AMD64%20FreeBSD%2010.x%20Shared%203.x/builds/240/steps/test/logs/stdio

Warning -- threading.\_dangling was modified by test_asyncio
  Before: \<_weakrefset.WeakSet object at 0x805f5d810\>
  After:  \<_weakrefset.WeakSet object at 0x807929948\>

**Comment by vstinner** [comment]
Xiang Zhang just fixed a bug to release resources in test_asyncio: bpo-30110, commit 981096f98b9c131594b0ac85ad01b63cbd11aa0a. No idea if it's enough to fix the warning.

**Comment by vstinner** [comment]
New changeset b9030674624c181d6e9047cdb14ad65bb6c84c66 by Victor Stinner in branch 'master':
bpo-30280: Cleanup threads in ayncio tests (bpo-2501)
https://github.com/python/cpython/commit/b9030674624c181d6e9047cdb14ad65bb6c84c66

**Comment by vstinner** [comment]
The commit b9030674624c181d6e9047cdb14ad65bb6c84c66 fixes one leaking thread. I don't know if it's enough.

Koobs gave me a SSH access to his FreeBSD buildbots (9, 10 and CURRENT), but I failed to reproduce the warning issue :-( I tried to run my system_load.py to run the tests when the system load is 5 or higher, but it doesn't help. I tried to run test_asyncio 4 times in parallel, but I was still unable to reproduce the bug.

The commit b9030674624c181d6e9047cdb14ad65bb6c84c66 not only fixes a bug (cleanup correctly one event loop), but it now also checks after \*each* test that we don't leak threads. At least, if the bug reoccurs and the test is run in verbose mode, we will get the name of the test which leaked threads!
